internal_ws/ykbh/src/lib.rs

```diff
@@ -139,37 +143,42 @@ impl LocalMem {
     }
 }
 
-/// A interpreter stack frame, containing allocated memory for the frames locals, and the function
+/// An interpreter stack frame, containing allocated memory for the frames locals, and the function
 /// symbol name and basic block index needed by the interpreter to continue interpreting after
 /// returning from a function call.
 struct StackFrame {
     /// Allocated memory holding live locals.
     mem: LocalMem,
-    /// The current basic block index of this frame. Upon returning from a function call it is used
+    /// The current basic block index of this frame. XXX i don't understand this sentence. Upon returning from a function call it is used
     /// to look up the previous basic block and check its terminator to decide where to continue
     /// interpreting.
     bbidx: ykpack::BasicBlockIndex,
-    /// Symbol name of this stack frame. Needed to retrieve the SIR body of the function which
+    /// woah, we malloc a `String` per stack frame?! that's going to be very slow. we need an alternative mechanism here. Symbol name of this stack frame. Needed to retrieve the SIR body of the function which
     /// contains the statements we want to interpret.
     func: String,
 }
 
-/// The SIR interpreter, also known as blackholing interpreter, is invoked when a guard fails in a
-/// trace. It is initalised with information from the trace, e.g. live variables, stack frames, and
-/// then run to get us back to a control point from where the normal interpreter can take over.
+/// The SIR interpreter, also known as the blackhole interpreter, is invoked when a guard fails in
+/// a trace. It is initialised with information from the trace (e.g. live variables, stack frames)
+/// and then runs until we are back to the control point, from where the normal interpreter can
+/// take over.
 pub struct SIRInterpreter {
-    /// Keeps track of active stack frames (most recent last).
+    /// Active stack frames (most recent last).
     frames: Vec<StackFrame>,
 }
 
 impl SIRInterpreter {
+    /// XXX i think this test is only needed for tests? if so, we should at least #[cfg(test)] it,
+    /// and probably give it a name that doesn't suggests it's the main way of initialising the
+    /// interpreter.
```

Comment:
Be aware that `#[cfg(test)]` [only works intra-crate](https://github.com/rust-lang/cargo/issues/8379). If this is tested from the external workspace, annotating it won't work and will lead to a confusing error message.

Reply:
Renamed to `from_symbol` in df14c2a.